### PR TITLE
Replace FFI to CUDAExtension for PyTorch 1.0.0

### PR DESCRIPTION
**Pull request by jinserk**
Hi @SeanNaren,
I know that PyTorch 1.0.0 includes brand new CTCLoss, however, my old codes were not being trained well with it. I've discussed about the issue with @t-vi [here](https://github.com/pytorch/pytorch/issues/12201), but it's still unclear that the bpc metric he mentioned could make the difference. So I modified the old ffi extension interface of this project to the new CppExtension and CUDAExtention of PyTorch 1.0.0 to keep it with my legacy codes. I also know that you are going to make this project deprecated. It's totally up to your decision to merge this PR. Thank you!

### CODE REVIEW

**Comment by SeanNaren** [comment]
oh this is awesome! I'll pull this in for sure, i'll give it a look over ASAP. But yes will be deprecating this for the built in loss function into pytorch, installing warp-ctc has been a history of pain for people :D

**Comment by jinserk** [comment]
It looks like the CI only checks cpu version pytorch, so I realize that the previous PR only works in gpu version. I've modified it to work on both of cpu and gpu versions. :D

**Comment by jinserk** [comment]
Hmm. the CI still failed. I guess the CI's pytorch is still the current stable 0.4.1 or so, not supporting the cpp extensions of pytorch.

**Comment by SeanNaren** [comment]
I'll keep this PR open till it becomes the stable GA release, that way I don't need to change the travis file to point to the nightly build!

**Comment by jerinphilip** [comment]
@jinserk: This works with 1.0.0, have you tested?

**Comment by SeanNaren** [comment]
Thanks Jinserk! I made a few changes/rebased down to one commit and got the tests working

**Comment by jinserk** [comment]
It's my big pleasure! Merry Christmas and Happy New Year to you!
